[CBRD-26271] Partition 생성시 복제 옵션 상속 및 승격 시 컬럼 속성 상속

youngjun9072 opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-26271>](http://jira.cubrid.org/browse/CBRD-26271)

### Purpose
HA 환경에서 파티션을 사용하기위해선 다음 두가지 사항이 만족되어야합니다.
1. 파티션 생성 시 자식 테이블에 복제 옵션이 추가되어야하며, 이는 부모 테이블의 복제 옵션을 따른다.
2. 파티션 승격 시 자식 테이블은 부모 테이블의 컬럼 속성 중 PK와 NOT NULL UK를 상속 받아야한다.
(현재는 컬럼속성을 상속받고 있지 않아 PK 혹은 NOT NULL UNIQUE가 제거된 복제 테이블이 생성 됩니다.)

이를 위해 테이블에 생성시 자식 테이블은 부모 테이블로부터 복제 옵션을 상속 받아야하며, 
파티션 승격시 자식 테이블은 부모 테이블의 컬럼 속성을 상속 받아야합니다.

### Implementation

파티션 생성
1. do_create_entity -> do_create_partition 호출
2. do_create_partition 에서 자식 클래스 생성 전 부모 클래스의 복제 옵션 조회
3. 자식 클래스에 플래그 설정

파티션 승격
1. do_promote_partition 함수에서 컬럼 속성 복사
2. 컬럼 속성 및 제약 초기화 루틴 삭제

### Remarks

코드상에 constraint를 복제하는 과정이 포함되어있고, 파라미터로 btid를 공유하게 되는데 
이에해대 문제점이 없을지 다음 두가지 방법으로 검증하였습니다. 

1. SHOW TRACE;를 이용한 index 스캔 여부
```
CREATE TABLE participant2 (
    host_year INT,
    nation CHAR(3),
    gold INT,
    silver INT,
    bronze INT
)
PARTITION BY RANGE (host_year) (
    PARTITION before_2000 VALUES LESS THAN (2000),
    PARTITION before_2008 VALUES LESS THAN (2008)
);

```
위와같은 테이블이 있을 때, host_year 를 일반 INT, PK, NOT NULL UNIQUE 세가지 CONSTRAINT를 반영해 테스트를 진행하였습니다.

결과 1-1: host_year INT
```
Query Plan:
  TABLE SCAN (dba.participant2__p__before_2000)
```
결과 1-2 host_year INT PRIMARY KEY
```
Query Plan:
  INDEX SCAN (dba.participant2__p__before_2000.pk_participant2_host_year) (key range: [dba.participant2__p__before_2000].host_year= ?:0 )
```
결과 1-3 host_year INT NOT NULL UNIQUE
```
Query Plan:
  INDEX SCAN (dba.participant2__p__before_2000.u_participant2_host_year) (key range: [dba.participant2__p__before_2000].host_year= ?:0 )
```


3. btree_delete_index 발생 시점
gdb 에서 btree_delete_index() 에 브레이크 포인트를 걸고 어느 시점에 발생하는지 관찰하였습니다. 
시나리오대로라면 프로모션시에는 호출되어선 안되고, 프로모션으로 인해 생성된 일반 테이블 삭제시에 호출되어야합니다.

예상대로 파티션된 테이블 생성 후 PROMOTE PARTITION 시에는 btree_delete_Index() 가 호출되지 않고,
삭제시에만 호출됩니다.

youngjun9072 committed:
Add replication attributes for partitions and inherit constraints on promotion

- On CREATE TABLE ... PARTITION BY, child partitions inherit the parent's replication option.
- On ALTER TABLE ... PROMOTE PARTITION, the child inherits the parent's attributes and constraints.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -4118,6 +4118,14 @@ do_create_partition (PARSER_CONTEXT * parser, PT_NODE * alter, SM_PARTITION_ALTE
 		}
 	    }
 
+	  error =
```
다음 처럼 복제 대상이 아닌지 미리 체크하는 로직이 더 가독성이 좋은 것 같습니다.

```
if (!sm_is_replication_class (pinfo->root_op))
  {
    ...
  }
```

youngjun9072 replied:
@hornetmj 반영하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7005,25 +7020,6 @@ do_promote_partition (SM_CLASS * class_)
 
-  /* Make sure we do not copy anything that actually belongs to the root class (the class to which this partition
-   * belongs to). This includes: auto_increment flags, unique indexes, primary keys, and foreign keys */
-  for (smattr = ctemplate->attributes; smattr != NULL; smattr = (SM_ATTRIBUTE *) smattr->header.next)
```
요구사항에 따라 promotion 시 포함되어야 하는 속성은 PK이며, 복제를 고려하여 RK 관련된 UK와 NON NULL 속성이 모두 존재하는 경우에 이를 유지하는 정도가 필요할 것 같습니다. 추가적으로, 이 속성들이 포함되었을 때 promotion 후 pk, uk 인덱스가 해당 테이블에 존재해야 하는데, 이 부분도 검증 요청드립니다.

youngjun9072 replied:
@hornetmj  반영하였습니다.

추가로 db_index를 출력해본 결과
```sql
CREATE TABLE participant3 (
    host_year INT PRIMARY KEY,
    nation CHAR(3),
    gold INT,
    silver INT,
    bronze INT
)
PARTITION BY RANGE (host_year) (
    PARTITION before_2000 VALUES LESS THAN (2000),
    PARTITION before_2008 VALUES LESS THAN (2008)
);
```
이 쿼리문으로 생성한 테이블에서 before_2000를 승격시킬 경우

```
<00001> index_name       : 'pk_participant3_host_year'
        is_unique        : 'YES'
        is_reverse       : 'NO'
        class_name       : 'participant3__p__before_2000'
        owner_name       : 'DBA'
        key_count        : 1
        is_primary_key   : 'YES'
        is_foreign_key   : 'NO'
        filter_expression: NULL
        have_function    : 'NO'
        comment          : NULL
        status           : 'NORMAL INDEX'
```
의 결과가 나옵니다. 또한, 아래와같이
```sql
CREATE TABLE participant3 (
    host_year INT NOT NULL UNIQUE,
    nation CHAR(3),
    gold INT,
    silver INT,
    bronze INT
)
PARTITION BY RANGE (host_year) (
    PARTITION before_2000 VALUES LESS THAN (2000),
    PARTITION before_2008 VALUES LESS THAN (2008)
);
```
host_year를 UK로 변경해 테이블을 생성할 경우
```
<00001> index_name       : 'u_participant3_host_year'
        is_unique        : 'YES'
        is_reverse       : 'NO'
        class_name       : 'participant3__p__before_2000'
        owner_name       : 'DBA'
        key_count        : 1
        is_primary_key   : 'NO'
        is_foreign_key   : 'NO'
        filter_expression: NULL
        have_function    : 'NO'
        comment          : NULL
        status           : 'NORMAL INDEX'
```
의 결과가 출력됩니다.

(host_year에 PK와 UK를 입력하지않을 경우 아무 인덱스가 출력되지 않습니다)

youngjun9072 replied:
@hornetmj 추가 테스트 결과 공유드립니다.
위의 테이블(PK, UK)에 대해 승격 후 다음과 같은 동일한 결과를 보입니다.
테스트 1은 부모 테이블로부터 인덱스가 별도로 반영되는게 맞는지 확인용이고,
테스트 2는 분리된 자식 테이블 내에 인덱스가 충돌나는지 확인용입니다.
```
// 테스트 1. 인덱스 분리 확인 테스트
insert into participant3 values (1990, 'kr', 1, 1, 1);
insert into participant3__p__before_2000 values(1990, 'kr', 1,1,1);

// 테스트 2. 인덱스 여부 테스트
insert into participant3__p__before_2000 values(1990, 'kr', 1,1,1);

In the command from line 1,

ERROR: Operation would have caused one or more unique constraint violations. INDEX pk_participant3_host_year(B+tree: 0|5952|5953) ON CLASS dba.participant3__p__before_2000(CLASS_OID: 0|209|5). key: 1990(OID: 0|6017|2).

```

youngjun9072 committed:
Apply review feedback
- Remove only the PK and NOT NULL UNIQUE reset logic during table promotion
- Improve code readability

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -4117,6 +4117,14 @@ do_create_partition (PARSER_CONTEXT * parser, PT_NODE * alter, SM_PARTITION_ALTE
 		  goto end_create;
 		}
 	    }
+	  if (!sm_is_replication_class (pinfo->root_op))
```
if와 else 블럭에서 모두 사용되기 때문에 위 reuse_oid나 tde_algo와 같은 변수를 사용하는게 좋을 것 같습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7004,22 +7021,14 @@ do_promote_partition (SM_CLASS * class_)
-	  smattr->flags &= ~(SM_ATTFLAG_NON_NULL);
-	  has_pk = true;
-	}
-      smattr->flags &= ~(SM_ATTFLAG_UNIQUE);
-      smattr->flags &= ~(SM_ATTFLAG_REVERSE_UNIQUE);
```
pk가 없는 경우에 대해서도 고려해야 할 것 같습니다. pk가 없고, uk만 있는 경우 not null 제약이 없기 때문에 promotion 시 제외 대상이 되어야 합니다.
1. pk 존재 - uk 인덱스는 항상 not null 보장 (uk 인덱스 생성 시 항상 pk 컬럼을 포함해야 하기 때문)
2. pk 존재 안함 - uk 인덱스는 not null 보장 안함
```
CREATE TABLE t (i INT) PARTITION BY LIST (i) (PARTITION p0 VALUES IN (1, 2), PARTITION p1 VALUES IN (3, 4));
CREATE UNIQUE INDEX idx2 ON t(i);
// desc t; 명령의 결과에서 Null 필드가 'YES' 확인
alter table t promote partition p1;
// uk 인덱스가 포함되도록 수정하면, null 값을 허용하는 상태
```

youngjun9072 replied:
@hornetmj 수정 반영했고, 예시에 적힌 sql실행 결과 노션으로 공유드렸습니다.

youngjun9072 committed:
commit

youngjun9072 committed:
commit

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into partition_replication

youngjun9072 committed:
temp

youngjun9072 committed:
Fix constraint inheritance logic

- Modified to inherit only PK and NOT NULL UNIQUE constraints
- Removed redundant flag reset code

youngjun9072 committed:
Fix bug where CONSTRAINT properties were reset when multiple non-NOT NULL UNIQUE columns existed

youngjun9072 committed:
Refactor constraint copy logic to handle only PK or NOT NULL UNIQUE constraints

- Modified the constraint copy process to replicate only PK and NOT NULL UNIQUE constraints.
- Extracted the related logic into a separate helper function for better clarity and maintainability.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7016,21 +7069,23 @@ do_promote_partition (SM_CLASS * class_)
       smattr->auto_increment = NULL;
       smattr->flags &= ~(SM_ATTFLAG_AUTO_INCREMENT);
-      if ((smattr->flags & SM_ATTFLAG_PRIMARY_KEY) != 0)
+      if (need_reset_notnull_unique_flags (smattr))
```
NIT:

검사 후 리셋하려는 대상이 uk이기 때문에 다음 네이밍이 더 적합해 보입니다. 또는 has_notnull_unique_constraints() 등을 추천합니다.
```suggestion
      if (need_reset_unique_flags (smattr))
```

youngjun9072 replied:
반영했습니다. has_notnull_unique_constraints() 로 변경되었습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7052,14 +7107,12 @@ do_promote_partition (SM_CLASS * class_)
   if (ctemplate->properties != NULL)
     {
-      if (has_pk)
+      if (!has_notnull_unique)
```
has_notnull_unique == true인 경우 제거되어야 할 속성(uk w/o not null)까지 제거되지 않을 것 같습니다.

추가적으로, 코드 수정 없이 다음 내용 TODO로 관리해 주세요.
- 파티션 승격 시 PK 제약을 함께 승격시키는 것이 요구사항이였고, 이 수정을 통해 반영 중 입니다.
- RK 후보인 unique not null 제약 역시 저의 요청으로 함께 승격시키는 작업을 이 수정에서 진행하셨습니다.
- 추후 unique not null의 경우 RK인 경우에만 함께 승격시키는 것이 더 적합할 것 같습니다.
- 위에서 말씀드린 것처럼 아직 RK가 고정되어 있지 않기 때문에 본 수정에서는 별도 처리는 하지 않으셔도 됩니다.

hornetmj commented on src/object/class_object.c:
```diff
@@ -536,6 +536,54 @@ classobj_map_constraint_to_property (SM_CONSTRAINT_TYPE constraint)
+ *   dest(in/out): destination attribute
+ */
+int
+classobj_copy_pk_and_notnull_unique_constraints (const SM_ATTRIBUTE * src, SM_ATTRIBUTE * dest)
```
ALTER PROMOTE 구문 외에도 CREATE TABLE 구문 등에서도 호출되는데, 각 케이스 별 검토가 완료된 상태인가요? 코드 리뷰 중 익숙하지 않은 코드부라 궁금해서 질문드립니다.

youngjun9072 replied:
@hornetmj 결론부터 말씀드리자면 CREATE TABLE시에도 classobj_init_attribute() 함수를 호출하는게 맞고, 2번 호출되지만 
2번 모두 classobj_copy_pk_and_notnull_unique_constraints() 를 호출하지는 않습니다.

classobj_copy_pk_and_notnull_unique_constraints() 의 호출 조건은 두가지입니다.
1. classobj_init_attribute(SM_ATTRIBUTE * src, SM_ATTRIBUTE * dest, int copy)에서 copy 값이 1이거나
2. classobj_init_attribute() 함수내에서 src->constraint의 값이 존재해야합니다.

CREATE TABLE 의 호출 루틴과 위의 조건을 요약하면
정리하면 처음  classobj_init_attribute(src, dest, copy) 호출시 아규먼트 copy=1 설정되지만 src->constraints의  값이 없고, 
이후 property와 constraint값이 채워진 후 classobj_init_attribute(src, dest, copy) 가 한번더 호출되지만 copy=0으로 설정되어 복사 루틴을 타지 않습니다.

반면 promote 시 이미 constraint 값이 있고(기존에 저장된 테이블을 참조하기때문), classobj_init_attribute()의 copy파라미터도 1로 설정되어 호출되기때문에 classobj_copy_pk_and_notnull_unique_constraints() 루틴을 탑니다.

P.S
분석하면서 한가지 고민은 flatten을 발생시킬 수 있는 루틴이 많은데 모든 케이스에 대해 역추적하거나(오늘도 일부 역추적을하다가 일단 포기한상태입니다...) copy, constraint 변수를 확인하기 어려운 상태입니다.
(루틴상으로는 do_drop, do_alter_one_clause_with_template()에서 PT_DROP_PARTITION, PT_COALESCE_PARTITION일 경우, sm_finish_class(), sm_update_class(), sm_update_class_with_auth() 등입니다. 
특히 sm_update_class() 가 상당히 많은곳에서 호출됩니다.) 
문제가 될것같진 않은데 추가 코멘트가 있으시면 남겨주시면 감사하겠습니다.

APPENDIX
아래는 do_create_entity에서 호출되는 로직 상세입니다.
1. do_create_entity -> dbt_finish_class() -> update_class() 
2. flatten_template() -> flatten_components() -> make_component_from_candidate() -> classobj_copy_attribute() -> classobj_init_attribute(src, dest, 1) 순서로 호출
3. 이 시점에서 CREATE TABLE로 생성된 template->attributes->constraints는 null 이기때문에 classobj_copy_pk_and_notnull_unique_constraints() 는 불리지 않습니다.
4. 이후 platten_components() 로 다시 돌아와 flatten_properties()로 properties 값 설정
5. 이후 update_class() 로 다시 돌아온 후 install_new_represantation() -> transfer_disk_structures() 에서 classobj_make_class_constraints() 를 통해 constraint 구조체 생성
6. install_new_represantation() -> classobj_install_template() -> classobj_init_attribute(src, dest, 0) 으로 copy값을 0으로 호출
과정을 거치게 됩니다.

youngjun9072 committed:
Refactor: Update function name and comments per review feedback

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -3984,6 +3985,7 @@ do_create_partition (PARSER_CONTEXT * parser, PT_NODE * alter, SM_PARTITION_ALTE
 
   reuse_oid = (smclass->flags & SM_CLASSFLAG_REUSE_OID) ? true : false;
   tde_algo = (TDE_ALGORITHM) smclass->tde_algorithm;
+  replication_opt = sm_is_replication_class (pinfo->root_op);
```
NIT:

SM_CLASS를 이미 획득했고, reuse_oid나 tde_algo 획득 방식과 일치

```suggestion
  replication_opt = (smclass->flags & SM_CLASSFLAG_DATA_REPLICATION_OFF) ? false : true;
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -6965,8 +7018,10 @@ do_promote_partition (SM_CLASS * class_)
   DB_CTMPL *ctemplate = NULL;
   SM_ATTRIBUTE *smattr = NULL;
-  bool has_pk = false;
+  bool has_notnull_unique = false;
```
```suggestion
  bool has_pk_or_uk_notnull = false;
```

youngjun9072 replied:
@hornetmj 모든 상황에서 pk는 상속이고, unique의 경우 not null unique만 상속하면되기때문에 has_notnull_unique가 나을것같습니다.

hornetmj replied:
@youngjun9072 문맥 설명 감사합니다. 확인했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -6952,6 +6971,40 @@ do_promote_partition_by_name (const char *class_name, const char *part_num, char
+ *    - Otherwise, return false and the UNIQUE-related flags will be reset.
+ */
+static bool
+has_notnull_unique_constraints (const SM_ATTRIBUTE * smattr)
```
```suggestion
has_pk_or_uk_notnull_constraints (const SM_ATTRIBUTE * smattr)
```

youngjun9072 replied:
@hornetmj has_notnull_unique와 동일한 이유로 PK는 모든 상황에서 기본 상속이라 not null unique만 판별하면 되는 함수입니다. 대신 has_notnull_unique_constraints() 함수에서 PK 판별이 포함된 SM_IS_CONSTRAINT_UNIQUE_FAMILY  대신 SM_CONSTRAINT_UNIQUE || SM_CONSTRAINT_REVERSE_UNIQUE 만 판별하도록 수정했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -6952,6 +6971,40 @@ do_promote_partition_by_name (const char *class_name, const char *part_num, char
+  bool has_unique = false;
+  SM_CONSTRAINT *c;
+
+  has_not_null = (smattr->flags & SM_ATTFLAG_NON_NULL) != 0;
```
has_not_null와 has_unique는 && 조건이기 때문에 여기서 has_not_null == false 라면, 아래 for 루프를 수행할 필요가 없습니다.

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -6952,6 +6971,40 @@ do_promote_partition_by_name (const char *class_name, const char *part_num, char
+ *    - Otherwise, return false and the UNIQUE-related flags will be reset.
+ */
+static bool
+has_notnull_unique_constraints (const SM_ATTRIBUTE * smattr)
```
has_notnull_unique_constraints() 함수명에 대해 두번째 제안드립니다. 사유는
1. 이 함수는 partition table의 제약사항 때문에 문제 없이 동작합니다. (이 제약들은 커멘트로 남겨주시면, SM_ATTRIBUTE를 개별적으로 검사해도 되는 이유가 명확해 질 것 같습니다.)
    - pk 또는 uk는 반드시 partition key여야 한다
    - partition key는 단일 컬럼만 허용 된다.
2. 1번의 제약사항이 없다면, 복합키에 걸린 제약을 검사해야 하기 때문에 SM_ATTRIBUTE 수준에서 판단할 수 없습니다.
3. 따라서, 이 함수가 파티션 테이블에 국한하여 사용되는 함수임을 명확하게 하기 위해 제안드립니다. 검토 후 의견주세요.
```suggestion
is_part_key_pk_or_uk_notnull (const SM_ATTRIBUTE * smattr)
```

youngjun9072 replied:
@hornetmj 위에서 언급한 함수의 이름과 동작이 정리되면 다시 검토해보려합니다.

hornetmj replied:
@youngjun9072 이 부분에 대해서도 검토 요청드립니다.

youngjun9072 replied:
@hornetmj 변수명을 has_notnull_unique 로 유지하기로하였고, 함수의 내용도 PK판별을 하지 않고 not null unique에 대해서만 수행중(PK는 언제나 상속합니다)이기때문에 함수명에 PK는 불필요해보입니다
다음 주석 내용을 추가하는건 어떨까요??
```
/* 파티션 테이블의 메타데이터 정합성을 유지하기 위해,
 * 루트 파티션 테이블에만 의미가 있는 속성들은 복사 시 제거하고,
 * 파티션에도 반드시 유지되어야 하는 제약은 그대로 보존합니다.
 *
 * EPIC CBRD-26096에 따른 전제 조건:
 *   - 파티션 키는 반드시 단일 컬럼이어야 하며,
 *   - 해당 컬럼은 PK 또는 NOT NULL UNIQUE로 정의되어야 합니다.
 *
 * 위 조건으로 인해 복합 키 여부를 별도로 검사할 필요가 없으며,
 * 단일 컬럼 단위(NOT NULL + UNIQUE) 판별만으로 파티션 키를 식별할 수 있습니다.
 *
 * 처리 규칙 요약:
 *   - PK(Primary Key)는 루트 테이블에서 정의되고 각 파티션으로 상속되므로
 *     여기에서 초기화하지 않습니다.
 *   - 반면, UNIQUE / REVERSE_UNIQUE 플래그는 루트 테이블의 메타데이터가
 *     그대로 복사되면 안 되므로, 파티션 키로 사용되는 단일 NOT NULL UNIQUE 컬럼
 *     한 개만 예외적으로 유지하고 나머지는 제거합니다.
 *   - FOREIGN KEY 및 PARTITION KEY 플래그 또한 루트 테이블 맥락 정보이므로
 *     파티션에서는 초기화합니다.
 */
```

hornetmj replied:
우선, 그대로 두시죠.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7016,21 +7071,23 @@ do_promote_partition (SM_CLASS * class_)
       smattr->auto_increment = NULL;
       smattr->flags &= ~(SM_ATTFLAG_AUTO_INCREMENT);
-      if ((smattr->flags & SM_ATTFLAG_PRIMARY_KEY) != 0)
+      if (has_notnull_unique_constraints (smattr))
```
```suggestion
      if (has_notnull_unique == false && has_notnull_unique_constraints (smattr) == true)
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -7052,14 +7109,12 @@ do_promote_partition (SM_CLASS * class_)
-	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_PRIMARY_KEY);
-	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_NOT_NULL);
+	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_REVERSE_UNIQUE);
+	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_UNIQUE);
```
NIT:
```suggestion
	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_UNIQUE);
	  classobj_drop_prop (ctemplate->properties, SM_PROPERTY_REVERSE_UNIQUE);
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/object/class_object.c:
```diff
@@ -536,6 +536,54 @@ classobj_map_constraint_to_property (SM_CONSTRAINT_TYPE constraint)
+ *   dest(in/out): destination attribute
+ */
+int
+classobj_copy_pk_and_notnull_unique_constraints (const SM_ATTRIBUTE * src, SM_ATTRIBUTE * dest)
```
```suggestion
static int
classobj_copy_pk_and_uk_notnull_constraints (const SM_ATTRIBUTE * src, SM_ATTRIBUTE * dest)
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj replied:
@youngjun9072 반영되어 있지 않습니다. 확인 요청드립니다.

youngjun9072 replied:
@hornetmj 함수명 변경이 누락되어있었습니다. 반영했습니다.

youngjun9072 committed:
Apply review feedback

- Performed minor refactoring and cleanup.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -6952,6 +6971,44 @@ do_promote_partition_by_name (const char *class_name, const char *part_num, char
+  has_not_null = (smattr->flags & SM_ATTFLAG_NON_NULL) != 0;
+  if (!has_not_null)
+    {
+      return FALSE;
```
NIT: bool 변수값 스타일 일관성
```suggestion
      return false;
```

youngjun9072 replied:
@hornetmj 반영했습니다.

youngjun9072 committed:
Add missing comments
